### ISSUE TITLE
(🐞) TypeGuard behaves inconsistently on methods.

### ISSUE THREAD

**Issue by sg495** [python/mypy#16618]
**Bug Report**

When type guards are implemented as methods, Mypy's type narrowing behaviour is inconsistent:

- It depends on whether the method only takes the implicit ``self`` argument or instead takes additional arguments.
- It further depends on whether the second argument has a default value, regardless of whether an explicit value is passed to it in the call expression.

This does not appear to be in compliance with [PEP 647](https://peps.python.org/pep-0647/), whose stated behaviour does not depend on any arguments further than the first explicit argument. Some relevant quotes from the PEP follow:

> Type checkers should assume that type narrowing should be applied to the expression that is passed as the first positional argument to a user-defined type guard. If the type guard function accepts more than one argument, no type narrowing is applied to those additional argument expressions.

> If the type guard function is implemented as an instance or class method, an implicit self or cls argument will also be passed to the function.

> If a type guard function is implemented as an instance method or class method, the first positional argument maps to the second parameter (after ``self`` or ``cls``).

>  A concern was raised that there may be cases where it is desired to apply the narrowing logic on ``self`` and ``cls``. [...] It was [...] decided that no special provision would be made for it. If narrowing of ``self`` or ``cls`` is required, the value can be passed as an explicit argument to a type guard function.

**Expected Behavior**

In Issue #16146, it was clarified that the intent of PEP 647 is for positionality to [depend on the call expression](https://github.com/python/mypy/issues/16146#issuecomment-1740229840): the first positional argument is ``c`` in the call expression ``a.f(c)``, but it is ``a`` in the call expression ``A.f(a, c)``.

Now, consider the following pattern where a type guard is implemented by an instance method, with arbitrary arguments:

```py
class A:
    def f(self, ...) -> TypeGuard[B]:
        #       ^^^ your choice of arguments here
        return isinstance(self, B)

class B(A): ...
```

The expectation is that call expressions in the form ``A.f(a, ...)``&mdash;where ``...`` here stands for arbitary values passed to other arguments in such a way as to make the call legal&mdash;will always result in type narrowing of ``a`` from ``a: A`` to ``a: B``. This behaviour should be independent of the number of additional arguments and/or whether additional arguments are given default values (mentioned here because default values increase the number of legal call expressions).

**Actual Behavior**

In the three code snippets below, Mypy fails to apply this logic to the case where methods only take the implicit argument ``self``, so that ``A.f(a)`` does not result in type narrowing for ``a`` while ``A.f(a, c)`` does. Furthermore, ``A.f(a, c)`` results in type narrowing behaviour for ``a`` when ``c`` is not given a default value, but fails to do so when a default value for ``c`` is specified, regardless of whether the call expression is ``A.f(a, c)`` (explicit value passed for ``c``) or ``A.f(a)`` (default value used for ``c``).

Pylance behaves as expected in all cases.

**To Reproduce**

❌ Case 1: only ``self`` argument.

```py
from __future__ import annotations
from typing import Any, TypeGuard, reveal_type

class A:
    def f(self) -> TypeGuard[B]:
        return isinstance(self, B)

class B(A): ...

a = A()

assert A.f(a)
reveal_type(a)
# Pylance: Type of "a" is "B"
# Mypy: Revealed type is "A"

b: B = a
# Mypy: Incompatible types in assignment (expression has type "A", variable has type "B")
# Pylance: OK
```

✅ Case 2: ``self`` argument plus one explicit argument. 

```py
from __future__ import annotations
from typing import TypeGuard, reveal_type

class A:
    def f(self, c: str) -> TypeGuard[B]:
        print(c)
        return isinstance(self, B)

class B(A): ...

a = A()
assert A.f(a, "Hi!")

reveal_type(a)
# Mypy: Revealed type is "B"
# Pylance: Type of "a" is "B"

b: B = a # No error
```

❌ Case 3: ``self`` argument plus one explicit argument with default value. Second argument passed explicitly in call expression.

```py
from __future__ import annotations
from typing import Any, TypeGuard, reveal_type

class A:
    def f(self, c: str = "") -> TypeGuard[B]:
        print(c)
        return isinstance(self, B)

class B(A): ...

a = A()
assert A.f(a, "Hi!")
reveal_type(a)
# Mypy: Revealed type is "A"
# Pylance: Type of "c" is "B"

b: B = a
# Mypy: Incompatible types in assignment (expression has type "A", variable has type "B")
# Pylance: OK
```

❌ Case 4: ``self`` argument plus one explicit argument with default value. Default value for second argument used implicitly in call expression.

```py
from __future__ import annotations
from typing import Any, TypeGuard, reveal_type

class A:
    def f(self, c: str = "") -> TypeGuard[B]:
        print(c)
        return isinstance(self, B)

class B(A): ...

a = A()
assert A.f(a)
reveal_type(a)
# Mypy: Revealed type is "A"
# Pylance: Type of "c" is "B"

b: B = a
# Mypy: Incompatible types in assignment (expression has type "A", variable has type "B")
# Pylance: OK
```

✅ Case 5: ``self`` argument plus two explicit arguments, only one of which has a default value. (Behaviour does not depend on whether the third argument is passed explicitly or not.)

```py
from __future__ import annotations
from typing import Any, TypeGuard, reveal_type

class A:
    def f(self, c: str, d: int = 0) -> TypeGuard[B]:
        print(c)
        return isinstance(self, B)

class B(A): ...

a = A()
assert A.f(a, "Hi")
reveal_type(a)
# Mypy: Revealed type is "B"
# Pylance: Type of "c" is "B"

b: B = a # No error
```

**Your Environment**

- Mypy version used: ``mypy 1.7.1 (compiled: yes)``
- Mypy command-line flags: ``--strict``
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: ``Python 3.12.0``

**Comment by erictraut** [comment]
You're using `TypeGuard` in a manner in which it wasn't intended in a few of these cases, so it's not surprising that you'll see inconsistent behaviors. If you use `TypeGuard` in the way it was intended, you will see correct and consistent behavior between type checkers. Note: I'm the author of PEP 647.

In "Case 1", you've defined an invalid type guard function because it has only a `self` parameter, but a user-defined type guard implemented as a method must have at least one parameter beyond `self`. Pyright emits an error indicating that a return type of `TypeGuard` cannot be used in this case. Mypy should perhaps emit a similar error here.

In "Case 2" through "Case 5", you've defined your type guard as an instance method, but you're calling it through a class binding rather than an instance binding and then explicitly passing an argument for `self`. I admittedly didn't consider this case when writing the PEP, nor (apparently) did any of the people who reviewed it. I would consider the behavior unspecified in this case. Don't do this if you want deterministic and consistent behavior across type checkers.

In "Case 6", you're using a user-defined type guard in a way that is consistent with its intended usage since it's defined as a class method.

In general, I recommend that you stick with functions or static methods if you want to define a user-defined type guard. As you've demonstrated above, using `TypeGuard` with instance and class methods can lead to ambiguous cases if they are used in ways that were not intended.

In any case, I don't consider mypy's behavior here to be out of compliance with the PEP.

If you see a compelling need for these edge cases to be clarified and the spec updated accordingly, you could make your case in the issue in the [Python Typing forum](https://discuss.python.org/c/typing/32).

**Comment by sg495** [comment]
The reason I consistently used the pattern ``A.f(a, ...)`` above is because it is the one considered [in your comment](https://github.com/python/mypy/issues/16146#issuecomment-1740229840) to Issue #16146:

> The PEP says that the expression passed as the first positional argument to the call should be narrowed. In your example, the first positional argument in the call expression ``A.f(a, c)`` is ``a.`` Mypy behaving in accordance with the text of the PEP. It's also working consistently with pyright, which is the reference implementation for PEP 647.

I agree with you in terms of what PEP 647 says and doesn't say: my main issue here is that Mypy's behaviour is inconsistent. This should be fixed, and I would prefer to see it go the Pylance way (allowing case 1) rather than the Pyright way (disallowing case 1). 

I will make my case for this on the forum, as you suggested: I have been frustrated again and again by the fact that ``self`` cannot be narrowed by type guards. I believe property-based and method-based type guards have natural use cases, which I would like to see implemented one day.

Case 1, specifically, was my last attempt at a work-around in this sense, before admitting defeat in the particular project I'm currently working on:

- It could be used as ``a.is_b()`` by users of my library who don't care about static typing, but who still need to check the properties that the typing ``a: B`` would enforce.
- It could also be used as ``A.is_b(a)`` by more sophisticated users who need the static type ``a: B`` recognised by their type-checker.
- Ideally, it would be used as a property ``a.is_b`` in both cases.
- The class/static method ``A.is_b(a)`` option has the disadvantage that ``A`` must be imported explicitly (or it has to be used as ``a.is_b(a)``, which would unfortunately make zero sense to my intended audience).

**Comment by KotlinIsland** [comment]
Try these out in [basedmypy](https://github.com/KotlinIsland/basedmypy), it handles type-guards in a much more intuitive way: 
- 🛝 [playground](https://mypy-play.net/?mypy=basedmypy-latest)
- 📚 [docs](https://kotlinisland.github.io/basedmypy/based_features.html#reinvented-type-guards)
